subcommand/server-acl-init/command_ent_test.go

```diff
@@ -0,0 +1,436 @@
+// +build enterprise
+
+package serveraclinit
+
+import (
+	"encoding/base64"
+	"fmt"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/sdk/testutil"
+	"github.com/mitchellh/cli"
+	"github.com/stretchr/testify/require"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes/fake"
+	"testing"
+)
+
+// Test that ACL policies get updated if namespaces config changes.
+func TestRun_ACLPolicyUpdates(t *testing.T) {
+	t.Parallel()
+	k8s, testAgent := completeEnterpriseSetup(t, resourcePrefix)
+	defer testAgent.Stop()
+	require := require.New(t)
+
+	ui := cli.NewMockUi()
+	firstRunArgs := []string{
+		"-server-label-selector=component=server,app=consul,release=" + releaseName,
+		"-resource-prefix=" + resourcePrefix,
+		"-k8s-namespace=" + ns,
+		"-create-client-token",
+		"-allow-dns",
+		"-create-mesh-gateway-token",
+		"-create-sync-token",
+		"-create-inject-namespace-token",
+		"-expected-replicas=1",
+		"-allow-dns",
+	}
+	// Our second run, we're going to update from namespaces disabled to
+	// namespaces enabled with a single destination ns.
+	secondRunArgs := append(firstRunArgs,
+		"-enable-namespaces",
+		"-consul-sync-destination-namespace=sync",
+		"-consul-inject-destination-namespace=dest")
+
+	// Run the command first to populate the policies.
+	cmd := Command{
+		UI:        ui,
+		clientset: k8s,
+	}
+	responseCode := cmd.Run(firstRunArgs)
+	require.Equal(0, responseCode, ui.ErrorWriter.String())
+
+	bootToken := getBootToken(t, k8s, resourcePrefix)
+	consul, err := api.NewClient(&api.Config{
+		Address: testAgent.HTTPAddr,
+		Token:   bootToken,
+	})
+	require.NoError(err)
+
+	// Check that the expected policies were created.
+	expectedPolicies := []string{
+		"dns-policy",
+		"client-token",
+		"catalog-sync-token",
+		"connect-inject-token",
+		"mesh-gateway-token",
+	}
+	policies, _, err := consul.ACL().PolicyList(nil)
+	require.NoError(err)
+	for _, expected := range expectedPolicies {
+		found := false
+		for _, p := range policies {
+			if p.Name == expected {
+				found = true
+
+				// We assert that the policy doesn't have any namespace config
+				// in it because later that's what we're using to test that it
+				// got updated.
+				policy, _, err := consul.ACL().PolicyRead(p.ID, nil)
+				require.NoError(err)
+				require.NotContains(policy.Rules, "namespace")
+				break
+			}
+		}
+		require.True(found, "Did not find "+expected)
+	}
+
+	// Re-run the command with namespace flags. The policies should be updated.
+	cmd = Command{
```

Comment:
Does this need to be defined again?

Reply:
Yes 😄found out about this the hard way. If we re-use the old command then old flags are still set. In this case it won't make a difference but if the first-run had set `-destination-consul-ns` then even though you set `-mirror-k8s` in the second run it would never take effect since the old flag would still be set.

In this case, it won't have effect but I think it's better to have it re-defined for two reasons:
1. If the test changes in the future it's very confusing to debug
1. Its a better representation of the real-world use-case. If a user upgrades with helm the command won't be re-used, it will be redefined since it's a brand new job. So we wouldn't want a side-effect in this case to cause the test to pass.

Reply:
done.